[CS-4083] Await for profile creation job (pooling)

oidoug opened the pull request:
### Description

This PR adds implementation for awaiting the profile creation process to finish, using a Lazy Query that starts pooling on first `jobQuery` call, and ends it once we get a successful response or an error.

UX will be improved once we have the loading / retry screen done in upcoming project. For now it requires users to wait until its done and to pull down to refresh in Wallet Screen to see the profile.

- [x] Completes #(CS-4083)

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS

### Update 2

- Using job-ticked-id again to strict pooling to a specific job;
- Showing Profile Screen after profile creation flow;
- Also shows non-blocking loading in Profile Screen while profile being created.

### Screenshots (updated)

https://user-images.githubusercontent.com/129619/183513213-e5110b69-c3f8-493e-ae02-c6998ca52467.mp4

oidoug committed:
Adds: hub-api query to job-ticket;
Lazy pooling until success response.
Hook to wait for success, updates flow to close purchase only after success.

oidoug committed:
Mock for useProfileJobAwait on usePurchaseProfile tests.

oidoug commented:
Updating it based on discussion.

oidoug committed:
Awaiting for profile improvements:
- On moving to wallet from profile purchase, a flag param tells assetlist to poll for profile;
- createProfileJob called with EOA, finds success state among returned array;
- cache invalidation is dispatched only after succesfull profile returns.

oidoug committed:
Removes unused prop.

oidoug commented:
PR updated, please take a look again @danibonilha and @paulinhapenedo

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -136,6 +138,22 @@ export const hubApi = createApi({
+          // makes a new query, as it should per documented here:
+          // https://redux-toolkit.js.org/rtk-query/usage/automated-refetching
+          dispatch(hubApi.util.invalidateTags([SafeCacheTags.SAFES]));
+        }
```
Does invalidating tags from different api works?  hubApi has its own tags, and safeApi have another, I'm not sure if they communicate. if it doesn't we should probably just trigger a simple refresh

oidoug replied:
That's probably why the invalidation it's not actually refreshing the safes. I can't find clear info in rtk-q docs about invalidating tags "across apis", so you're probably right. I'll remove this invalidation then.

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -136,6 +138,22 @@ export const hubApi = createApi({
       }),
       transformResponse: ({ data }) => data?.attributes?.did,
     }),
+    createProfileJob: builder.query<string, { eoa: string }>({
```
create, gives a impression that is some kind of mutation, you could call it, getProfileJobStatus

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -136,6 +138,22 @@ export const hubApi = createApi({
+    createProfileJob: builder.query<string, { eoa: string }>({
+      query: ({ eoa }) => `${routes.profileInfo.jobTicket}?eoa=${eoa}`,
+      transformResponse: ({ data }) =>
+        data.find((job: any) => job.attributes?.state === 'success'),
```
This assumes the only job ticket existing is related to the profile, which might not be true, we should filter by the jobType if this query is related just with profile.

oidoug replied:
True, I forgot about the possibility of having more job types.

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -136,6 +138,22 @@ export const hubApi = createApi({
+    createProfileJob: builder.query<string, { eoa: string }>({
+      query: ({ eoa }) => `${routes.profileInfo.jobTicket}?eoa=${eoa}`,
+      transformResponse: ({ data }) =>
+        data.find((job: any) => job.attributes?.state === 'success'),
```
it would be nice to add the typing of job, this way we don't need to go back and forth on the api to take a look

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -136,6 +138,22 @@ export const hubApi = createApi({
+    createProfileJob: builder.query<string, { eoa: string }>({
+      query: ({ eoa }) => `${routes.profileInfo.jobTicket}?eoa=${eoa}`,
+      transformResponse: ({ data }) =>
+        data.find((job: any) => job.attributes?.state === 'success'),
```
Also this find, will return when it finds the first successful job-ticket, and if there's a pending on in the end of the array, it won't be accounted for

oidoug replied:
This query is only considering successful jobs now. If we intend on using the pending jobs too, using a filter by job-type would be better. Until we make use of pending jobs, do you think getting the first successful is enough?

danibonilha commented on cardstack/src/components/AssetList/useAssetList.ts:
```diff
@@ -106,12 +109,34 @@ export const useAssetList = ({
+
+  // shouldAwaitForProfile will be set if screen was open from profile creation flow.
+  useEffect(() => {
+    if (params?.shouldAwaitForProfile && accountAddress) {
```
I'm not sure we need this nav param, it might be less confusing to just trigger the job ticket at start, if we change the screen in the end of the profile flow, for example, to go to Profile, this jobPolling will never be triggered. I believe the `useProfileJobAwait` should be smart enough to handle everything, in the case we want to change it's place, the only change would be to call it in the expected screen

oidoug replied:
Yes, I considered and tested doing that, but it requires a separate global state to keep track when to start pooling. I did that to avoid having a slice with just that flag, but might be necessary anyway.

oidoug committed:
Using job-ticked-id again to strict pooling to a specific job;
Showing Profile Screen after profile creation flow;
Shows non-blocking loading in Profile Screen while profile being created.

oidoug committed:
Moving loading flag to polling hook.

oidoug commented:
Known issue: on Profile tab, while loading the new safe, you see the "create a profile" state for a second.

I'm investigating but prefer to handle this in a new PR.

paulinhapenedo commented on cardstack/src/hooks/usePurchaseProfile.ts:
```diff
@@ -59,12 +59,14 @@ export const usePurchaseProfile = (profile: CreateProfileInfoParams) => {
               Alert({
                 title: 'Success',
-                message:
-                  'Your profile will appear shortly, pull to refresh in a minute or so',
+                message: 'Your profile will appear shortly',
```
nit: move these into a strings file.

oidoug replied:
This alert will be removed.

danibonilha commented on cardstack/src/hooks/usePurchaseProfile.ts:
```diff
@@ -59,12 +59,14 @@ export const usePurchaseProfile = (profile: CreateProfileInfoParams) => {
-                    onPress: () => navigate(Routes.WALLET_SCREEN),
+                    onPress: () =>
+                      navigate(Routes.PROFILE_SCREEN, {
+                        profileCreationJobID: profileJobInfo?.included[0]?.id,
```
nit: we could transform the query response, and return the job id

oidoug replied:
yeah, makes sense.

danibonilha commented on cardstack/src/screens/ProfileScreen/useProfileScreen.ts:
```diff
@@ -0,0 +1,83 @@
+    isUninitialized,
+  } = usePrimarySafe();
+
+  // Profile creation job takes time to complete, this hook keeps pooling until the profile is ready.
```
This comment is duped

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -46,7 +49,7 @@ enum CacheTag {
   reducerPath: 'hubApi',
   baseQuery: fetchHubBaseQuery,
-  tagTypes: [...Object.values(CacheTag)],
+  tagTypes: [...Object.values(CacheTag), ...Object.values(SafeCacheTags)],
```
We should drop this change

danibonilha commented on cardstack/src/services/hub/hub-types.ts:
```diff
@@ -77,3 +77,17 @@ export interface PostProfilePurchaseQueryResult {
+    };
+  };
+}
+
```
nice

danibonilha commented on cardstack/src/hooks/__tests__/usePurchaseProfile.test.ts:
```diff
@@ -45,6 +45,14 @@ jest.mock('@cardstack/services', () => ({
+    error: undefined,
+  })),
+}));
+
```
This seems a leftover

oidoug replied:
Yes, thanks for pointing that out.

danibonilha commented on cardstack/src/screens/ProfileScreen/useProfileScreen.ts:
```diff
@@ -0,0 +1,83 @@
+
+  // When the new profile is done, we need to refresh the safes list.
+  useEffect(() => {
+    if (data?.attributes?.state === 'success') {
```
nit: instead of returning the data, we could return something like, isJobCompleted/Successful, and handle whatever we want, OR the `useProfileJobPolling` could receive a `onJobCompleted` callback then we wouldn't need this useEffect here

oidoug replied:
Yes, thats a good idea.

I was considering we would use the response data when pending somehow in the future, but we can refactor if needed later.

danibonilha commented on cardstack/src/screens/ProfileScreen/ProfileScreen.tsx:
```diff
@@ -9,31 +8,24 @@ import {
-
-  const isRefreshingForNewAccount = useIsFetchingDataNewAccount(isFetching);
+    redirectToSwitchNetwork,
+  } = useProfileScreen();
```
nice job on creating the screen hook

danibonilha commented on cardstack/src/hooks/usePurchaseProfile.ts:
```diff
@@ -59,12 +59,14 @@ export const usePurchaseProfile = (profile: CreateProfileInfoParams) => {
 
               Alert({
```
Wondering if we still need this alert, since this flow will be redirected to a loading screen

oidoug replied:
I agree, there are too many alerts in this process currently, even the one for a successful purchase might not be needed.

For now, I'll remove this one only.

oidoug committed:
Improvements based on PR requests.

oidoug committed:
Adds missing callback call

oidoug committed:
Fixes type for error

danibonilha reviewed (approved):
LGTM! 🚀
